reference scripts not always executed

JaredCorduan opened the pull request:
Two functions in the ledger code were making the faulty assumption that every reference script is executed:

* The function `collectTwoPhaseScriptInputs` was making this faulty assumption when collecting cost models. We would not be able to get away with fixing this were it not for the fact that Alonzo has always had a V1 cost model, and we can depend on it always having one.
* The function `evaluateTransactionExecutionUnits` (which is used by the CLI for building transactions) was making this faulty assumption when collecting needed transaction contexts. This resulted in outputs containing Plutus V1 reference scripts being unspendable by the CLI (even as a regular input, ie not just reference inputs). To fix this problem, I factored out `knownToNotBe1Phase` out of `collectTwoPhaseScriptInputs`.

resolves #2851

Commit:
do not require cost model for all ref inputs

We do not need to require the presence of a cost model for all languages
present in reference scripts, but rather only those that are actually
needed.

Commit:
evaluateTransactionExecutionUnits used languages

The helper function evaluateTransactionExecutionUnits for the CLI was
assuming that every refenence script inside of every reference input
corresponded to a script which needed a transaction context.
This rendered outputs with Plutus V1 reference scripts unspendable in
the CLI.

Jimbo4350 commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -156,24 +175,16 @@ collectTwoPhaseScriptInputs ::
-        Just (AlonzoScript.PlutusScript lang script) -> Just (sp, lang, script)
-        Just (AlonzoScript.TimelockScript _) -> Nothing
-        Nothing -> Nothing
+    needed = mapMaybe (knownToNotBe1Phase scriptsAvailable) $ scriptsNeeded utxo tx
```
Instead of `needed` what about `onlyPlutusScripts`? For me `needed` does not immediately tell me about the kind of scripts needed or what `needed` is. I would have to dig further into how `needed` was created.

JaredCorduan replied:
yea, that's better. is `neededAndConfirmedToBePlutus` too long? :grimacing:

Jimbo4350 replied:
That's fine as well

Jimbo4350 commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs:
```diff
@@ -175,15 +169,16 @@ evaluateTransactionExecutionUnits pp tx utxo ei sysS costModels = do
-    scripts = txscripts utxo tx
+    scriptsAvailable = txscripts utxo tx
     needed = scriptsNeeded utxo tx
+    nonPhase1Needed = mapMaybe (knownToNotBe1Phase scriptsAvailable) needed
```
Same thing here `onlyPlutusScripts`. `nonPhase1Needed` is confusing to me.

JaredCorduan replied:
I was trying to stress that we might not have _all_ the plutus scripts. we've just gathered the ones we can, based on what we know. `onlyPlutusScripts` is mathematically correct, but maybe elides the presence of sneaky scripts we can't identify?

Jimbo4350 replied:
> sneaky scripts we can't identify

Which sneaky scripts? Doesn't `txscripts` get all the possible scripts our transaction is using?

JaredCorduan replied:
yep, `txscripts` gathers all the scripts that we have available for executing (both simple and Plutus). Some of which we don't want to execute, because they are not actually needed. The ones we need are identified by a collection of hashes, namely `needed`. In order to figure out which of the needed ones are Plutus scripts, we look them up in `scriptsAvailable`.  The sneaky part is that there could be a script hash that corresponds to a script that we need, but that we cannot identify as being a Plutus script since it is not in `scriptsAvailable`.

JaredCorduan replied:
I've gone with `neededAndConfirmedToBePlutus` for this variable and the one above that we talked about.

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -132,6 +133,24 @@ instance (CC.Crypto crypto) => FromCBOR (CollectError crypto) where
       dec 3 = SumD BadTranslation <! From
       dec n = Invalid n
 
+-- Given an script purpose and a script hash, determine if it is *not*
```
```suggestion
-- Given a script purpose and a script hash, determine if it is *not*
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -132,6 +133,24 @@ instance (CC.Crypto crypto) => FromCBOR (CollectError crypto) where
+    Just (AlonzoScript.PlutusScript lang script) -> Just (sp, lang, script)
+    Just (AlonzoScript.TimelockScript _) -> Nothing
+    Nothing -> Nothing
+
```
Just a suggestion:
```suggestion
knownToNotBe1Phase scriptsAvailable (sp, sh) = do
  AlonzoScript.PlutusScript lang script <- sh `Map.lookup` scriptsAvailable
  Just (sp, lang, script)

```

JaredCorduan replied:
much nicer!

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:
```diff
@@ -156,24 +175,22 @@ collectTwoPhaseScriptInputs ::
    in case missingCMs of
         l : _ -> Left [NoCostModel l]
-        _ -> merge (apply costModels) (map redeemer needed) (map getscript needed) (Right [])
+        _ ->
```
I just realized the whole thing should be made more efficient:
```haskell
  let usedLanguages = Set.fromList [lang | (_, lang, _) <- neededAndConfirmedToBePlutus]
      costModels = unCostModels $ getField @"_costmdls" pp
      missingCMs = Set.filter (`Map.notMember` costModels) usedLanguages
   in case Set.lookupMin missingCMs of
        Just l -> Left [NoCostModel l]
        Nothing -> ...
```

JaredCorduan replied:
thanks!

JaredCorduan replied:
done

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs:
```diff
@@ -162,7 +156,7 @@ evaluateTransactionExecutionUnits pp tx utxo ei sysS costModels = do
       let getInfo :: Language -> Either TranslationError (Language, VersionedTxInfo)
           getInfo lang = (,) lang <$> txInfo pp lang ei sysS utxo tx
-          txInfos = map getInfo (Set.toList $ languagesUsed (Map.elems scripts))
+          txInfos = [getInfo lang | (_, lang, _) <- neededAndConfirmedToBePlutus]
```
This will force `getInfo` to be called as many times as there are scripts, instead of as many times as there are languages. In order to fix this we need to nub it:
```suggestion
          txInfos = map getInfo $ Set.toList $
                      Set.fromList [lang | (_, lang, _) <- neededAndConfirmedToBePlutus]
```

JaredCorduan replied:
fantastic!

lehins reviewed (changes requested):
There are couple of suggestions that fix performance. Otherwise looks good.

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs:
```diff
@@ -129,24 +123,26 @@ evaluateTransactionExecutionUnits pp tx utxo ei sysS costModels =
         pure $
           Map.mapWithKey
-            (findAndCount pp (array (PlutusV1, PlutusV2) ctx))
+            (findAndCount pp (array (minBound :: Language, maxBound) ctx))
```
I like that. I thought about suggesting it too ;)

lehins replied:
This actually won't work, because if `ctx` is missing a `Language` this will be bottom.

lehins reviewed (approved):
Looks great!

Commit:
fix dangerous use of Arary.!

findAndCount (used in a CLI helper function) was using Array.! in a
place where is was not guaranteed to succeed. We now check the bounds.

Additionally, one of the errors had a misleading name.
RedeemerNotNeeded was actually accouting for a case when a redeemer
pointed to a script hash which we could not resolve.

lehins committed:
Fix partial usage of `Array` by switching to `Map`

lehins committed:
Fix spelling in the type constructor and rename to `TransactionScriptFailure`

Also made contents strict.

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs:
```diff
@@ -54,40 +54,40 @@ import qualified Plutus.V1.Ledger.Api as PV1
 
 -- | Script failures that can be returned by 'evaluateTransactionExecutionUnits'.
-data ScriptFailure c
+data TransactionScriptFailure c
```
:tada:
